proj/hps_accel/gateware/test_filter_store.py

```diff
@@ -0,0 +1,194 @@
+                yield self.dut.input.payload.eq(filter_value)
+                yield
+            yield self.dut.input.valid.eq(0)
+            yield from self.read_out()
+        self.run_sim(process)
+
+    def test_it_can_tolerate_delays_while_reading(self):
+        def process():
+            yield from self.reset_and_fill()
+            for i in range(0, len(self.filter_values), 4):
+                yield self.dut.output[0].ready.eq(0)
+                yield self.dut.output[1].ready.eq(0)
+                yield self.dut.output[2].ready.eq(0)
+                yield self.dut.output[3].ready.eq(0)
+                yield self.dut.next.eq(0)
+                # Delay before consuming each value
+                # (Delaying for an odd number of cycles is more likely to catch
+                # bugs, since the store contains an even number of elements)
+                for _ in range(13):
+                    yield
+                self.assertEqual((yield self.dut.output[0].valid), 1)
+                self.assertEqual((yield self.dut.output[1].valid), 1)
+                self.assertEqual((yield self.dut.output[2].valid), 1)
+                self.assertEqual((yield self.dut.output[3].valid), 1)
+                self.assertEqual((yield self.dut.output[0].payload), self.filter_values[i + 0])
+                self.assertEqual((yield self.dut.output[1].payload), self.filter_values[i + 1])
+                self.assertEqual((yield self.dut.output[2].payload), self.filter_values[i + 2])
+                self.assertEqual((yield self.dut.output[3].payload), self.filter_values[i + 3])
+                yield self.dut.output[0].ready.eq(1)
+                yield self.dut.output[1].ready.eq(1)
+                yield self.dut.output[2].ready.eq(1)
+                yield self.dut.output[3].ready.eq(1)
+                yield self.dut.next.eq(1)
+                yield
+        self.run_sim(process)
+
+    def test_it_can_be_reset_while_filling(self):
+        def process():
+            # Reset and partially fill some elements
+            yield self.dut.num_words.payload.eq(len(self.filter_values))
+            yield self.dut.num_words.valid.eq(1)
+            yield
+            yield self.dut.num_words.valid.eq(0)
+            self.assertEqual((yield self.dut.input.ready), 1)
+            yield self.dut.input.payload.eq(99)
+            yield self.dut.input.valid.eq(1)
+            yield
+            self.assertEqual((yield self.dut.input.ready), 1)
+            yield self.dut.input.payload.eq(99)
+            yield
+            # Now reset and fill again from the start
+            yield from self.reset_and_fill()
+            yield from self.read_out()
+        self.run_sim(process)
+
+    # TODO(dcallagh): fix reset logic so that this case passes.
+    # It currently breaks because the first output values get stuck at their
+    # output sources after the first fill, and remain stuck there during the
+    # second fill and up to the subsequent read.
```

Comment:
I believe the root cause of the bug is in SinglePortMemory. I have a rewrite in progress to simplify that class in order to more easily fix the bug.